test/common/buffer/owned_impl_test.cc

```diff
@@ -340,6 +340,69 @@ TEST_P(OwnedImplTest, ReserveCommit) {
   }
 }
 
+TEST_P(OwnedImplTest, ReserveCommitReuse) {
+  Buffer::OwnedImpl buffer;
+  verifyImplementation(buffer);
+
+  static constexpr uint64_t NumIovecs = 2;
+  Buffer::RawSlice iovecs[NumIovecs];
+
+  // Reserve 8KB and commit all but a few bytes of it, to ensure that
+  // the last slice of the buffer can hold part but not all of the
+  // next reservation. Note that the buffer implementation might
+  // allocate more than the requested 8KB. In case the implementation
+  // uses a power-of-two allocator, the subsequent reservations all
+  // request 16KB.
+  uint64_t num_reserved = buffer.reserve(8192, iovecs, NumIovecs);
+  EXPECT_EQ(1, num_reserved);
+  iovecs[0].len_ = 8000;
+  buffer.commit(iovecs, 1);
+  EXPECT_EQ(8000, buffer.length());
+
+  // Reserve 16KB. The resulting reservation should span 2 slices.
+  // Commit part of the first slice and none of the second slice.
+  num_reserved = buffer.reserve(16384, iovecs, NumIovecs);
+  EXPECT_EQ(2, num_reserved);
+  const void* first_slice = iovecs[0].mem_;
+  const void* second_slice = iovecs[1].mem_;
+  iovecs[0].len_ = 1;
+  buffer.commit(iovecs, 1);
+  EXPECT_EQ(8001, buffer.length());
+
+  // Reserve 16KB again, and check whether we get back the uncommitted
+  // second slice from the previous reservation.
+  num_reserved = buffer.reserve(16384, iovecs, NumIovecs);
+  EXPECT_EQ(2, num_reserved);
+  EXPECT_EQ(static_cast<const uint8_t*>(first_slice) + 1,
+            static_cast<const uint8_t*>(iovecs[0].mem_));
+  EXPECT_EQ(second_slice, iovecs[1].mem_);
+}
+
+TEST_P(OwnedImplTest, ReserveReuse) {
+  Buffer::OwnedImpl buffer;
+  verifyImplementation(buffer);
+
+  static constexpr uint64_t NumIovecs = 2;
+  Buffer::RawSlice iovecs[NumIovecs];
+
+  // Reserve some space and leave it uncommitted.
+  uint64_t num_reserved = buffer.reserve(8192, iovecs, NumIovecs);
+  EXPECT_EQ(1, num_reserved);
+  const void* first_slice = iovecs[0].mem_;
+
+  // Reserve more space and verify that it begins with the same slice from the last reservation.
+  num_reserved = buffer.reserve(16384, iovecs, NumIovecs);
+  EXPECT_EQ(2, num_reserved);
+  EXPECT_EQ(first_slice, iovecs[0].mem_);
+  const void* second_slice = iovecs[1].mem_;
+
+  // Repeat the last reservation and verify that it yields the same slices.
+  num_reserved = buffer.reserve(16384, iovecs, NumIovecs);
+  EXPECT_EQ(2, num_reserved);
+  EXPECT_EQ(first_slice, iovecs[0].mem_);
+  EXPECT_EQ(second_slice, iovecs[1].mem_);
+}
```

Comment:
Can you make a test that confirms the memory problem is resolved, by using `Memory::Stats::totalCurrentlyAllocated()` and/or `Memory::Stats::totalCurrentlyReseved()`?

Ideally this test should fail without your source/... changes, and pass with them.

See (for example) test/integration/stats_integration_test.cc, test ClusterMemoryTestRunner.MemoryLargeClusterSizeWithStats. Note that this test only makes sense to run with TCMALLOC, hence the early exit in that test method based on `Stats::TestUtil::hasDeterministicMallocStats()`.

Reply:
Let me know if that test doesn't make sense @brian-pane . It contains some shell classes that are needed to create two independent runs.

Reply:
This PR already contains a test case to confirm the memory problem is resolved: `ReserveCommitReuse` in `test/common/buffer/owned_impl_test.cc`. I recommend sticking with that test approach, rather than the indirectness of an integration test that looks at memory usage for the whole process.

Reply:
Let's definitely keep your unit tests, but I'm unclear why the hesitation about subtracting off after-before memory; it's very stable and the overhead from the rest of the process is not relevant.

I wouldn't necessarily have gone down this route but experience so far with the stat memory tests indicates that these sorts of tests appear to be are consistent and deterministic enough.

And as the PR description does claim to fix #6617 I think this is the most direct way to test that the leak is gone.